src/main/java/com/fasterxml/jackson/core/JsonGenerator.java

```diff
@@ -1978,12 +1979,13 @@ public WritableTypeId writeTypePrefix(WritableTypeId typeIdDef) throws IOExcepti
         } else {
             // No native type id; write wrappers
             // Normally we only support String type ids (non-String reserved for native type ids)
-            String idStr = (id instanceof String) ? (String) id : String.valueOf(id);
+            String idStr = (id instanceof String) ? (String) id : Objects.toString(id, null);
             typeIdDef.wrapperWritten = true;
 
             Inclusion incl = typeIdDef.include;
             // first: can not output "as property" if value not Object; if so, must do "as array"
             if ((valueShape != JsonToken.START_OBJECT)
+                    && idStr != null 
```

Comment:
I guess I don't quite understand logic here: change will prevent changing of output for some cases from non-Array to Array. But not prevent output or anything like that.

So... what use case being resolved by this change?

Reply:
Seeing your comment in the PR comment I assume that you have a good grasp of the idea for change. I continue commenting there as find that context more relevant.